katdal/categorical.py

```diff
@@ -17,107 +17,93 @@
+      - It ensures that array-valued sensor values become properly comparable
+        (avoiding array-valued booleans resulting from standard comparisons).
+
+    The former is needed because :class:`SensorData` is treated as a structured
+    array even if it contains object values. The latter is needed because the
+    equality operator crops up in hard-to-reach places like inside list.index().
 
     Parameters
     ----------
-    hashable_value : object that is hashable
-        The wrapped version of a sensor value - if not available, use static
-        method :meth:`wrap` instead (current implementation expects a pickle)
+    value : object
+        The sensor value to be wrapped
 
     Note
     ----
-    The current implementation uses pickle to make generic objects hashable,
-    even though it is not recommended - see e.g. this `blog post
-    <http://www.aminus.org/blogs/index.php/2007/11/03/pickle_dumps_not_suitable_for_hashing>`_.
-    The main problem is with unordered data like dicts and sets, where the item
-    order in the pickle string may differ even though the objects are equal;
-
-      pickle.dumps({1: 0, 9: 0}) == pickle.dumps({9: 0, 1: 0})
-
-    evaluates to False, even though the dicts are equal. On top of this there
-    is also the issue of potentially differing pickle protocols.
+    The object is explicitly non-hashable to force comparisons to be by value
+    instead of by identity for ndarrays.
 
     """
 
-    def __init__(self, hashable_value):
-        self.hashable_value = hashable_value
+    def __init__(self, value):
+        self.unwrapped = value
 
     def __repr__(self):
         """Short human-friendly string representation of wrapper object."""
         return "<katdal.%s { %r } at 0x%x>" % \
-               (self.__class__.__name__, self.unwrap(), id(self))
+               (self.__class__.__name__, self.unwrapped, id(self))
 
     def __str__(self):
         """Longer human-friendly string representation of wrapped object."""
-        return str(self.unwrap())
+        return str(self.unwrapped)
 
     def __eq__(self, other):
         """Equality comparison operator."""
-        if not isinstance(other, HashableValueWrapper):
-            other = HashableValueWrapper.wrap(other)
-        return self.hashable_value == other.hashable_value
+        if isinstance(other, ComparableArrayWrapper):
+            other = other.unwrapped
+        if isinstance(self.unwrapped, np.ndarray) or isinstance(other, np.ndarray):
+            return np.array_equal(self.unwrapped, other)
+        else:
+            return self.unwrapped == other
 
     def __ne__(self, other):
         """Inequality comparison operator."""
         return not self == other
 
     def __lt__(self, other):
         """Less-than comparison operator."""
-        raise TypeError("HashableValueWrapper objects are considered unorderable")
+        return self.unwrapped < other
```

Comment:
Shouldn't this be `self.unwrapped < other.unwrapped`, at least if `other` is a wrapper?